src/include/avro_copy.hpp

```diff
@@ -0,0 +1,163 @@
+#pragma once
+
+#include "duckdb/function/copy_function.hpp"
+#include <avro.h>
+#include "duckdb/common/serializer/memory_stream.hpp"
+#include "duckdb/common/file_system.hpp"
+
+namespace duckdb {
+
+struct AvroCopyFunction {
+	static CopyFunction Create();
+};
+
+struct WriteAvroBindData : public FunctionData {
+public:
+	WriteAvroBindData(CopyFunctionBindInput &input, const vector<string> &names, const vector<LogicalType> &types);
+	WriteAvroBindData() {
+	}
+	virtual ~WriteAvroBindData();
+
+public:
+	unique_ptr<FunctionData> Copy() const override {
+		auto res = make_uniq<WriteAvroBindData>();
+		res->names = names;
+		res->types = types;
+
+		res->schema = avro_schema_incref(schema);
+		res->json_schema = json_schema;
+
+		res->interface = avro_value_iface_incref(interface);
+		return res;
+	}
+
+	bool Equals(const FunctionData &other_p) const override {
+		auto &other = other_p.Cast<WriteAvroBindData>();
+		if (interface != other.interface) {
+			//! FIXME: is comparing pointers too much?
+			return false;
+		}
+		if (other.types.size() != types.size()) {
+			return false;
+		}
+		if (other.names.size() != names.size()) {
+			return false;
+		}
+		if (json_schema != other.json_schema) {
+			return false;
+		}
+		return true;
+	}
+
+public:
+	vector<string> names;
+	vector<LogicalType> types;
+
+	//! The schema of the file to write
+	avro_schema_t schema = nullptr;
+	string json_schema;
+
+	//! The interface through which new avro values are created
+	avro_value_iface_t *interface = nullptr;
+};
+
+struct AvroInMemoryBuffer {
+public:
+	AvroInMemoryBuffer(Allocator &allocator, idx_t initial_capacity = 0)
+	    : allocator(allocator), capacity(initial_capacity) {
+		if (initial_capacity) {
+			Allocate(capacity);
+		}
+	}
+	~AvroInMemoryBuffer() {
+		if (data) {
+			allocator.FreeData(data, capacity);
+		}
+	}
+
+public:
+	void Resize(idx_t new_capacity) {
+		D_ASSERT(this->capacity < new_capacity);
+		if (data) {
+			//! Free the old buffer directly, we don't need to copy the old contents
+			allocator.FreeData(data, this->capacity);
+			data = nullptr;
+		}
+		Allocate(new_capacity);
+	}
+	void ResizeAndCopy(idx_t new_capacity) {
+		data_ptr_t old_data = data;
+		auto old_capacity = capacity;
+
+		data = nullptr;
+		Allocate(new_capacity);
+		memcpy(data, old_data, old_capacity);
+		allocator.FreeData(old_data, old_capacity);
+	}
+	data_ptr_t GetData() {
+		return data;
+	}
+	idx_t GetCapacity() const {
+		return capacity;
+	}
+
+private:
+	void Allocate(idx_t new_capacity) {
+		D_ASSERT(!data);
+		data = allocator.AllocateData(new_capacity);
+		capacity = new_capacity;
+	}
+
+public:
+	Allocator &allocator;
```

Comment:
Should these fields be public? If this becomes a class, maybe they could be a friend of the `WriteAvroGlobalState` class?

Reply:
Yea, could have been private, nothing here is accessed, I am however not addressing this